pyomo/repn/plugins/standard_form.py

```diff
@@ -0,0 +1,537 @@
+    Block,
+    Objective,
+    Constraint,
+    Var,
+    Param,
+    Expression,
+    SortComponents,
+    Suffix,
+    SymbolMap,
+    maximize,
+)
+from pyomo.opt import WriterFactory
+from pyomo.repn.linear import LinearRepnVisitor
+from pyomo.repn.util import (
+    FileDeterminism,
+    FileDeterminism_to_SortComponents,
+    categorize_valid_components,
+    initialize_var_map_from_column_order,
+    ordered_active_constraints,
+)
+
+### FIXME: Remove the following as soon as non-active components no
+### longer report active==True
+from pyomo.core.base import Set, RangeSet, ExternalFunction
+from pyomo.network import Port
+
+logger = logging.getLogger(__name__)
+
+RowEntry = collections.namedtuple('RowEntry', ['constraint', 'bound_type'])
+
+
+# TODO: make a proper base class
+class LinearStandardFormInfo(object):
+    """Return type for LinearStandardFormCompiler.write()
+
+    Attributes
+    ----------
+    c : scipy.sparse.csr_array
+
+        The objective coefficients.  Note that this is a sparse array
+        and may contain multiple rows (for multiobjective problems).  The
+        objectives may be calculated by "c @ x"
+
+    A : scipy.sparse.csc_array
+
+        The constraint coefficients.  The constraint bodies may be
+        calculated by "A @ x"
+
+    rhs : numpy.ndarray
+
+        The constraint right-hand sides.
+
+    rows : List[Tuple[_ConstraintData, int]]
+
+        The list of Pyomo constraint objects corresponding to the rows
+        in `A`.  Each element in the list is a 2-tuple of
+        (_ConstraintData, row_multiplier).  The `row_multiplier` will be
+        +/- 1 (indicating if the row was multiplied by -1 (corresponding
+        to a constraint lower bound or +1 (upper bound).
+
+    columns : List[_VarData]
+
+        The list of Pyomo variable objects corresponding to columns in
+        the `A` and `c` matricies.
+
+    eliminated_vars: List[Tuple[_VarData, NumericExpression]]
+
+        The list of variables from the original model that do not appear
+        in the standard form (usually because they were replaced by
+        nonnegative variables).  Each entry is a 2-tuple of
+        (:py:class:`_VarData`, :py:class`NumericExpression`|`float`).
+        The list is in the necessary order for correct evaluation (i.e.,
+        all variables appearing in the expression must either have
+        appeared in the standard form, or appear *earlier* in this list.
+
+    """
+
+    def __init__(self, c, A, rhs, rows, columns, eliminated_vars):
+        self.c = c
+        self.A = A
+        self.rhs = rhs
+        self.rows = rows
+        self.columns = columns
+        self.eliminated_vars = eliminated_vars
+
+    @property
+    def x(self):
+        return self.columns
+
+    @property
+    def b(self):
+        return self.rhs
+
+
+@WriterFactory.register(
+    'compile_standard_form', 'Compile an LP to standard form (`min cTx s.t. Ax <= b`)'
+)
+class LinearStandardFormCompiler(object):
```

Comment:
I originally went with "compile" because things like named constants (e.g., Params) are substituted / evaluated and simplified out -- as are all expressions.  "Convert" starts to overlap with the "`pyomo convert` machinery that converts models to other files (i.e., a wrapper around writers).  I could be convinced to change the name.

Reply:
Hmmmm, I see. I don't seem to have any better ideas at the moment, so let's go with it!